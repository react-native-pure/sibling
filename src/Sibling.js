import * as React from "react"
import SiblingEmitter, {SiblingEvents} from "./SiblingEmitter";

class Sibling {
    constructor() {
        this.id = 0;
    }

    append(element: React.Element) {
        const id = this.id;
        SiblingEmitter.emit(SiblingEvents.append, {
            id: id,
            node: React.cloneElement(element, {
                key: `sibling-${id}`
            })
        });
        this.id++;
        return function () {
            SiblingEmitter.emit(SiblingEvents.remove, id);
        };
    }
}

export default new Sibling();