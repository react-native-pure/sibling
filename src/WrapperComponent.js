import * as React from "react";
import {View, StyleSheet} from "react-native"
import Siblings from "./Siblings";

const styles = StyleSheet.create({
    container: {
        flex: 1
    }
})

function WrapperComponent(props) {
    return (
        <View style={styles.container}>
            {props.children}
            <Siblings/>
        </View>
    );
}

export default React.memo(WrapperComponent);