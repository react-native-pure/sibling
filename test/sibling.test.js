import * as React from "react"
import {Text} from "react-native"
import renderer from 'react-test-renderer'
import WrapperComponent from "../src/WrapperComponent"
import Sibling from "../src/Sibling"

test("mount <WrapperComponent/>", () => {
    const component = renderer.create(
        <WrapperComponent/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test(`mount <WrapperComponent/> with children`, () => {
    const component = renderer.create(
        <WrapperComponent>
            <Text>abc</Text>
        </WrapperComponent>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test(`append/remove Sibling element`, () => {
    const component = renderer.create(
        <WrapperComponent>
            <Text>Append Sibling Element</Text>
        </WrapperComponent>
    );
    const remove = Sibling.append(
        <Text>I am a sibling element</Text>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    remove();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test(`update sibling element`, () => {
    class Test extends React.Component {
        state = {
            message: ""
        };

        render() {
            return <Text>{this.state.message}</Text>
        }

        hello() {
            this.setState({message: "hello"});
        }
    }

    let testRef = null;

    const component = renderer.create(
        <WrapperComponent>
            <Test ref={ref => testRef = ref}></Test>
        </WrapperComponent>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    testRef.hello();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test(`can not append when unmount`, () => {
    const component = renderer.create(
        <WrapperComponent>
        </WrapperComponent>
    );
    component.unmount();
    Sibling.append(
        <Text>hello</Text>
    );
    expect(component.toJSON()).toMatchSnapshot();
});