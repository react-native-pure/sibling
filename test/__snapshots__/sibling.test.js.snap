// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`append/remove Sibling element 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    Append Sibling Element
  </Text>
  <Text>
    I am a sibling element
  </Text>
</View>
`;

exports[`append/remove Sibling element 2`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    Append Sibling Element
  </Text>
</View>
`;

exports[`can not append when unmount 1`] = `null`;

exports[`mount <WrapperComponent/> 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`mount <WrapperComponent/> with children 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    abc
  </Text>
</View>
`;

exports[`update sibling element 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    
  </Text>
</View>
`;

exports[`update sibling element 2`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    hello
  </Text>
</View>
`;
